### PR TITLE
feat:debug operator at local

### PR DESCRIPTION
**Pull request by jiuker**
Now:
in ui, We can use `kubectl proxy` at local
in controller, we can start it, but it can't connect k8s's endpoint at all.
in sidecar, it start with minio. We can't do this.

- [x] support svc forward
- [x] support headness forward
- [x] support domain once
- [x] enable kubeconfig
- [x] todo:reopen this
- [x] goimport
- [ ] tls
- [ ] update https://github.com/minio/operator/blob/master/CREDITS with license of new entry
- [ ] rename `dev` package

How to test it:
export KUBECONFIG="kubeconfigPath"
export DEVELOPMENT=true

args:
controller --kubeconfig=kubeconfigPath

### CODE REVIEW

File: pkg/controller/dev/portforward/forward.go [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+		clientSet, err := kubernetes.NewForConfig(cfg)
+		if err != nil {
+			panic(err)
+		}
+		GlobalDebugConfig.clientSet = clientSet
+	}
+}
+
+// PortForwarder creates a new PortForwarder using kubectl tooling
+func PortForwarder(
```
Comment (shtripat) [inline pkg/controller/dev/portforward/forward.go:74]: Better if function names are verbs

File: pkg/controller/dev/portforward/forward.go [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+}
+
+var podIPv4Regex = regexp.MustCompile(`^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$`)
+
+func portForwarder(
+	ctx context.Context,
+	httpReq *http.Request,
+) (string, error) {
+	if podIPv4Regex.MatchString(httpReq.URL.Host) {
+		return "", fmt.Errorf("IPV4 %s,Don't need forward", httpReq.URL.Host)
```
Comment (shtripat) [inline pkg/controller/dev/portforward/forward.go:118]: Minor: space after `,`

File: pkg/controller/dev/portforward/forward.go [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+			return "", err
+		}
+
+		var podName string
+		var ports []string
+
+		// we test one
+		for _, s := range endPoint.Subsets {
+			for _, addr := range s.Addresses {
+				if addr.TargetRef.Kind == "Pod" {
```
Comment (shtripat) [inline pkg/controller/dev/portforward/forward.go:166]: Can we join these two if conditions with `&&` ?

**Comment by jiuker** [comment]
@shtripat I need to know if this can work out of K8S.If yes we can fix code detail.

File: pkg/controller/dev/portforward/forward.go [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+	hostTargetMutex map[string]*sync.Mutex
+	clientSet       *kubernetes.Clientset
+	cfg             *rest.Config
+}
+
+// GlobalDebugConfig is for global debuginfo , set it once.
+var GlobalDebugConfig = DebugConfig{}
+
+func init() {
+	Kubeconfig := os.Getenv("KUBECONFIG")
```
Comment (shtripat) [inline pkg/controller/dev/portforward/forward.go:56]: Can we use `github.com/minio/pkg/env` for getting env vars?

File: go.mod [hunk 5]
```diff
@@ -130,6 +130,7 @@ require (
 	github.com/minio/md5-simd v1.1.2 // indirect
 	github.com/minio/sha256-simd v1.0.0 // indirect
 	github.com/mitchellh/mapstructure v1.4.3 // indirect
+	github.com/moby/spdystream v0.2.0 // indirect
```
Comment (shtripat) [inline go.mod:133]: Do we need to update https://github.com/minio/operator/blob/master/CREDITS with license of new entry here?

File: pkg/controller/dev/portforward/forward.go [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License, version 3,
+// along with this program.  If not, see <http://www.gnu.org/licenses/>
+
+package portforward
```
[the remark this answers is not in the document]
Comment (jiuker) [reply pkg/controller/dev/portforward/forward.go:15]: I will rename it if it works.

**Comment by jiuker** [comment]
tls is big problem. Will continue when I have thought.

**Comment by dvaldivia** [comment]
I think this PR is adding to much to achieve a task that is simple to archive by using the common rest config so controller can also go through `kubectl proxy`

The way I usually develop locally is by modifying `pkg/controller/controller.go` lines `80 to 84` I comment them and replace them for 
```go
	var err error
	cfg := &rest.Config{
		Host:            "http://localhost:8001",
		TLSClientConfig: rest.TLSClientConfig{Insecure: true},
		APIPath:         "/",
		BearerToken:     "eyJhbGciOiJSUzI1NiIsImtpZCI6InFETTJ6R21jMS1NRVpTOER0SnUwdVg1Q05XeDZLV2NKVTdMUnlsZWtUa28ifQ.eyJpc3MiOiJrdWJlcm5ldGVzL3NlcnZpY2VhY2NvdW50Iiwia3ViZXJuZXRlcy5pby9zZXJ2aWNlYWNjb3VudC9uYW1lc3BhY2UiOiJkZWZhdWx0Iiwia3ViZXJuZXRlcy5pby9zZXJ2aWNlYWNjb3VudC9zZWNyZXQubmFtZSI6ImRldi1zYS10b2tlbi14eGxuaiIsImt1YmVybmV0ZXMuaW8vc2VydmljZWFjY291bnQvc2VydmljZS1hY2NvdW50Lm5hbWUiOiJkZXYtc2EiLCJrdWJlcm5ldGVzLmlvL3NlcnZpY2VhY2NvdW50L3NlcnZpY2UtYWNjb3VudC51aWQiOiJmZDVhMzRjNy0wZTkwLTQxNTctYmY0Zi02Yjg4MzIwYWIzMDgiLCJzdWIiOiJzeXN0ZW06c2VydmljZWFjY291bnQ6ZGVmYXVsdDpkZXYtc2EifQ.woZ6Bmkkw-BMV-_UX0Y-S_Lkb6H9zqKZX2aNhyy7valbYIZfIzrDqJYWV9q2SwCP20jBfdsDS40nDcMnHJPE5jZHkTajAV6eAnoq4EspRqORtLGFnVV-JR-okxtvhhQpsw5MdZacJk36ED6Hg8If5uTOF7VF5r70dP7WYBMFiZ3HSlJBnbu7QoTKFmbJ1MafsTQ2RBA37IJPkqi3OHvPadTux6UdMI8LlY7bLkZkaryYR36kwIzSqsYgsnefmm4eZkZzpCeyS9scm9lPjeyQTyCAhftlxfw8m_fsV0EDhmybZCjgJi4R49leJYkHdpnCSkubj87kJAbGMwvLhMhFFQ",
	}
```

this makes it so the controller starts and attempts to connect to kubernetes via port 8001, the bearer token really does nothing.

Perhaps it's better to introduce a simple `OPERATOR_DEV_MODE` flag to switch between `InClusterConfig` and manual `rest.Config`, we don't need to introduce proxying capabilities imo

**Pull request closed without merging**